### PR TITLE
Support IN operator for LIST

### LINKED ISSUE #11158: Syntax error near ? when use prepared statement with operator 'in'
### What happens?

prepare statement doesn't allow parameter appears after operator `in`. For example, my query is:

```python
sql = "select * from bars_cache_status where symbol in ? and frame_type=?"
cursor.execute(sql, (symbols, 6))
```
will raise error:

图片

if I use this one:

```python
sql = "select * from bars_cache_status where symbol in ('000001.SZ', '000002.SZ') and frame_type=?"
cursor.execute(sql, (symbols, 6))
```
duckdb will be happy.

However, symbols maybe a very large list, so I'm afraid I can't hardcode it into sql clause.

So is this a feature not supported yet, or just I don't know how to use `in` operator and prepare statement together?

### To Reproduce

1. create the following table:
```bash
duckdb test.duckdb
```
```sql
CREATE SEQUENCE IF NOT EXISTS bars_cache_status_id;
create table if not exists bars_cache_status (
    "id" INT default nextval('bars_cache_status_id') primary key,
    "start" date,
    "stop" date,
    "symbol" string, --同步的证券代码
    "frame_type" int  --记录属于哪个周期。1分钟和1天允许不一样的起点
);

insert into bars_cache_status (start, stop, SYMBOL, frame_type) values (
    '2022-01-01', '2024-01-01', '000001.SZ', 6);
insert into bars_cache_status (start, stop, SYMBOL, frame_type) values (
    '2022-01-01', '2024-01-01', '000004.SZ', 6);
```
2. execute the following code:
```python
import duckdb
con = duckdb.connect("test.duckdb")

symbols = ("000001.SZ", "000002.sz")

with con.cursor() as cursor:
    sql = "select * from bars_cache_status where symbol in ? and frame_type=?"
    query = cursor.execute(sql, (symbols, 6))
    records = query.fetchnumpy()

```

### OS:

windows

### DuckDB Version:

0.10.0

### DuckDB Client:

Python

### Full Name:

Aaron Yang

### Affiliation:

NA

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a release build (and could not test with a nightly build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes https://github.com/duckdb/duckdb/issues/11158

These now work as expected:

```sql
SELECT 2 IN [1,2,3] -- true
```

```sql
create table tbl as select [1,2,3] a;

select 2 IN a from tbl; -- true
```

```sql
PREPARE v1 AS SELECT 1 IN ?

EXECUTE v1([1, 2, 3]) -- true
```

### CODE REVIEW

File: test/sql/function/generic/test_in.test [hunk 1]
```diff
@@ -139,3 +139,75 @@ NULL	NULL
...
+----
+false
+
+query I
+SELECT 1 IN ((1, 2, 3));
+----
+true
+
+query I
+SELECT 1 IN ((2, 2, 3));
```
Comment (Mytherin) [inline test/sql/function/generic/test_in.test:158]: Could we add a test for `SELECT 1 IN col FROM tbl`?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR!

Perhaps we can limit the scope of the PR to `x IN list` for now? I'm not sure what the semantics for `x IN struct` or `x IN map` should be. In Python this would check the keys - but it looks like you are checking the values.

Can we also make this work for columns and expressions instead of only for literals?

**Comment by Tishj** [comment]
> In Python this would check the keys - but it looks like you are checking the values.

I'm checking the keys as well

**Comment by Tishj** [comment]
> x IN struct

That is only meant to be used for unnamed structs (tuples), but you're right that's not very well defined or tested, I'll add more unhappy path testing

**Comment by Tishj** [comment]
@Mytherin can you restart <https://github.com/duckdb/duckdb/actions/runs/9889303153/job/27320569664?pr=12920> ?
It looks like a spurious failure

File: test/sql/function/generic/test_in.test [hunk 2]
```diff
@@ -139,3 +139,147 @@ NULL	NULL
...
+
+query I
+SELECT 1 IN [2, 2, 3];
+----
+false
+
+statement ok
+PREPARE v1 AS SELECT 1 IN ?
+
+mode skip
```
Comment (Mytherin) [inline test/sql/function/generic/test_in.test:155]: Is this left in on purpose?

File: test/sql/function/generic/test_in.test [hunk 3]
```diff
@@ -139,3 +139,147 @@ NULL	NULL
...
+NULL	NULL
+1	1
+2	1
+3	1
+
+query II
+with list_provider as (
+	select [1, 2, 3, 4, 5, 6, 7, 8] as lst
+)
+SELECT i, i NOT IN lst FROM (
```
Comment (Mytherin) [inline test/sql/function/generic/test_in.test:261]: Can we modify this test so `NOT IN` returns true for a member?

File: src/function/scalar/string/contains.cpp [hunk 4]
```diff
@@ -151,15 +152,39 @@ struct ContainsOperator {
...
+	auto list_fun = GetListFunction();
+	ScalarFunctionSet set("contains");
+	set.AddFunction(string_fun);
+	set.AddFunction(list_fun);
+	return set;
+}
+
+ScalarFunction ContainsFun::GetStringContains(ClientContext &context) {
+	auto set = ContainsFun::GetFunctions();
+	return set.GetFunctionByArguments(context, {LogicalType::VARCHAR, LogicalType::VARCHAR});
```
Comment (Mytherin) [inline src/function/scalar/string/contains.cpp:182]: Binding for this seems excessive - can we move the `ScalarFunction string_fun` into this function and instead call it from within `ContainsFun::GetFunctions`? That way we also don't need a client context here.

File: src/include/duckdb/function/scalar/list/contains_or_position.hpp [hunk 5]
```diff
@@ -81,6 +87,48 @@ static void TemplatedContainsOrPosition(DataChunk &args, Vector &result, bool is
 	}
 }
 
+template <LogicalTypeId RETURN_TYPE>
+unique_ptr<FunctionData> ListContainsOrPositionBind(ClientContext &context, ScalarFunction &bound_function,
```
Comment (Mytherin) [inline src/include/duckdb/function/scalar/list/contains_or_position.hpp:91]: Does this need to be moved to the header? Can't we just call `ListContainsFun::GetFunction` in `contains` to add it to the function set?

**Review by Mytherin — COMMENTED** [review]
Thanks for the changes! LGTM. Some more feedback below. I think the changeset can be made smaller as I think a lot of the function movement can be avoided.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**